### PR TITLE
[ACCSEC-21632] Add test to AESCipher

### PR DESCRIPTION
**Pull request by SergioFierro**
## Jira Ticket
- ACCSEC-21632

## Description
- Add unit and instrumented tests to AESCipher

## Commit message
- Add test to AESCipher

## Testing
- [ ] Added unit tests
- [ ] Ran unit tests successfully
- [ ] Added documentation for public APIs and/or Wiki

### CODE REVIEW

File: security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt [hunk 1]
```diff
@@ -3,25 +3,28 @@
  */
 package com.twilio.security.crypto
 
+import com.twilio.security.crypto.key.authentication.Authenticator
```
Comment (AuthyApps) [inline security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt:6]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt#L6"></span>Imports must be ordered in lexicographic order without any empty lines in-between</td>
    </tr>
  </tbody>
</table>

File: security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt [hunk 2]
```diff
@@ -147,6 +150,77 @@ class AESGCMNoPaddingCipherTests {
     assertNotEquals(encryptedData2, encryptedData3)
   }
 
+  @Test
+  fun testEncrypt_withCipherAndSuccessfulAuthentication_shouldReturnEncryptedData() {
+    val data = "message".toByteArray()
+    val authenticator = TestAuthenticator()
+    val template = AESGCMNoPaddingCipherTemplate(alias).templateForCreation()
+    val cipher = androidKeyManager.cipher(template)
+    cipher.encrypt(data, authenticator, { encryptedData ->
```
Comment (AuthyApps) [inline security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt:159]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt#L159"></span>Missing newline after "("</td>
    </tr>
  </tbody>
</table>

Comment (AuthyApps) [inline security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt:159]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt#L159"></span>Missing newline after ","</td>
    </tr>
  </tbody>
</table>

File: security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt [hunk 3]
```diff
@@ -147,6 +150,77 @@ class AESGCMNoPaddingCipherTests {
...
+      val decrypted = cipher.decrypt(encryptedData)
+      assertTrue(data.contentEquals(decrypted))
+    }, { fail() })
```
Comment (AuthyApps) [inline security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt:162]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt#L162"></span>Missing newline after ","</td>
    </tr>
  </tbody>
</table>

Comment (AuthyApps) [inline security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt:162]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt#L162"></span>Missing newline before ")"</td>
    </tr>
  </tbody>
</table>

File: security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt [hunk 4]
```diff
@@ -147,6 +150,77 @@ class AESGCMNoPaddingCipherTests {
...
+  }
+
+  @Test
+  fun testEncrypt_withCipherAndFailedAuthentication_shouldReturnError() {
+    val data = "message".toByteArray()
+    val expectedError = RuntimeException()
+    val authenticator = TestAuthenticator(expectedError)
+    val template = AESGCMNoPaddingCipherTemplate(alias).templateForCreation()
+    val cipher = androidKeyManager.cipher(template)
+    cipher.encrypt(data, authenticator, { fail() }, { error ->
```
Comment (AuthyApps) [inline security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt:172]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt#L172"></span>Missing newline after "("</td>
    </tr>
  </tbody>
</table>

Comment (AuthyApps) [inline security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt:172]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt#L172"></span>Missing newline after ","</td>
    </tr>
  </tbody>
</table>

File: security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt [hunk 5]
```diff
@@ -147,6 +150,77 @@ class AESGCMNoPaddingCipherTests {
...
+      assertEquals(expectedError, error)
+    })
```
Comment (AuthyApps) [inline security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt:174]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt#L174"></span>Missing newline before ")"</td>
    </tr>
  </tbody>
</table>

File: security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt [hunk 6]
```diff
@@ -147,6 +150,77 @@ class AESGCMNoPaddingCipherTests {
...
+        EncryptedData(
+          AlgorithmParametersSpec(
+            parameters.encoded, parameters.provider.name, parameters.algorithm
+          ),
+          doFinal(data)
+        )
+      }
+    val cipher = androidKeyManager.cipher(template)
+    assertNotNull((cipher as? AESCipher)?.key)
+    cipher.decrypt(encryptedData, authenticator, { decryptedData ->
```
Comment (AuthyApps) [inline security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt:195]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt#L195"></span>Missing newline after "("</td>
    </tr>
  </tbody>
</table>

Comment (AuthyApps) [inline security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt:195]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt#L195"></span>Missing newline after ","</td>
    </tr>
  </tbody>
</table>

File: security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt [hunk 7]
```diff
@@ -147,6 +150,77 @@ class AESGCMNoPaddingCipherTests {
...
+      assertTrue(data.contentEquals(decryptedData))
+    }, { fail() })
```
Comment (AuthyApps) [inline security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt:197]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt#L197"></span>Missing newline after ","</td>
    </tr>
  </tbody>
</table>

Comment (AuthyApps) [inline security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt:197]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt#L197"></span>Missing newline before ")"</td>
    </tr>
  </tbody>
</table>

File: security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt [hunk 8]
```diff
@@ -147,6 +150,77 @@ class AESGCMNoPaddingCipherTests {
...
+        EncryptedData(
+          AlgorithmParametersSpec(
+            parameters.encoded, parameters.provider.name, parameters.algorithm
+          ),
+          doFinal(data)
+        )
+      }
+    val cipher = androidKeyManager.cipher(template)
+    assertNotNull((cipher as? AESCipher)?.key)
+    cipher.decrypt(encryptedData, authenticator, { fail() }, { error ->
```
Comment (AuthyApps) [inline security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt:219]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt#L219"></span>Missing newline after "("</td>
    </tr>
  </tbody>
</table>

Comment (AuthyApps) [inline security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt:219]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt#L219"></span>Missing newline after ","</td>
    </tr>
  </tbody>
</table>

File: security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt [hunk 9]
```diff
@@ -147,6 +150,77 @@ class AESGCMNoPaddingCipherTests {
...
+      assertEquals(expectedError, error)
+    })
```
Comment (AuthyApps) [inline security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt:221]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/androidTest/java/com/twilio/security/crypto/AESGCMNoPaddingCipherTests.kt#L221"></span>Missing newline before ")"</td>
    </tr>
  </tbody>
</table>

File: security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt [hunk 10]
```diff
@@ -86,6 +88,102 @@ class AESCipherTest {
...
+    whenever(androidKeyStoreOperations.getCipherForEncryption(eq(cipherAlgorithm), eq(key))).thenReturn(
+      cipher
+    )
+    argumentCaptor<(Cipher) -> Unit>().apply {
+      whenever(authenticator.startAuthentication(eq(cipher), capture(), any())).then {
+        firstValue.invoke(cipher)
+      }
+    }
+    whenever(androidKeyStoreOperations.encrypt(eq(data), eq(cipher))).thenReturn(expectedEncryptedData)
+    aesCipher.encrypt(data, authenticator, { encryptedData ->
```
Comment (AuthyApps) [inline security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt:106]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt#L106"></span>Missing newline after "("</td>
    </tr>
  </tbody>
</table>

Comment (AuthyApps) [inline security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt:106]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt#L106"></span>Missing newline after ","</td>
    </tr>
  </tbody>
</table>

File: security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt [hunk 11]
```diff
@@ -86,6 +88,102 @@ class AESCipherTest {
...
+      assertEquals(expectedEncryptedData, encryptedData)
+    }, { fail() })
```
Comment (AuthyApps) [inline security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt:108]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt#L108"></span>Missing newline after ","</td>
    </tr>
  </tbody>
</table>

Comment (AuthyApps) [inline security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt:108]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt#L108"></span>Missing newline before ")"</td>
    </tr>
  </tbody>
</table>

File: security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt [hunk 12]
```diff
@@ -86,6 +88,102 @@ class AESCipherTest {
...
+    whenever(androidKeyStoreOperations.getCipherForDecryption(eq(cipherAlgorithm), eq(key), eq(encryptedData))).thenReturn(
+      cipher
+    )
+    argumentCaptor<(Cipher) -> Unit>().apply {
+      whenever(authenticator.startAuthentication(eq(cipher), capture(), any())).then {
+        firstValue.invoke(cipher)
+      }
+    }
+    whenever(androidKeyStoreOperations.decrypt(eq(encryptedData), eq(cipher))).thenReturn(expectedData)
+    aesCipher.decrypt(encryptedData, authenticator, { decryptedData ->
```
Comment (AuthyApps) [inline security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt:154]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt#L154"></span>Missing newline after "("</td>
    </tr>
  </tbody>
</table>

Comment (AuthyApps) [inline security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt:154]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt#L154"></span>Missing newline after ","</td>
    </tr>
  </tbody>
</table>

File: security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt [hunk 13]
```diff
@@ -86,6 +88,102 @@ class AESCipherTest {
...
+      assertEquals(expectedData, decryptedData)
+    }, { fail() })
```
Comment (AuthyApps) [inline security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt:156]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt#L156"></span>Missing newline after ","</td>
    </tr>
  </tbody>
</table>

Comment (AuthyApps) [inline security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt:156]: <table data-meta="generated_by_Linter">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/twilio/twilio-verify-android/blob/acfefe4ef52888b000994497481b7b030ae4ed08/security/src/test/java/com/twilio/security/crypto/key/cipher/AESCipherTest.kt#L156"></span>Missing newline before ")"</td>
    </tr>
  </tbody>
</table>

**Comment by AuthyApps** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Fix code convention problems</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_Linter">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by AlejandroOrozco — APPROVED** [review]
LGTM 👍

**Comment by AuthyApps** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Fix code convention problems</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_Linter">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by yafuquen — APPROVED** [review]

**Pull request merged**